packages/core/src/aad.ts

```diff
@@ -0,0 +1,253 @@
+// Copyright 2020 Cognite AS
+
+import {
+  PublicClientApplication,
+  SilentRequest,
+  AuthenticationResult,
+  Configuration,
+  AccountInfo,
+  InteractionRequiredAuthError,
+  EndSessionRequest,
+  RedirectRequest,
+  PopupRequest,
+  LogLevel,
+  BrowserSystemOptions,
+  CacheOptions,
+} from '@azure/msal-browser';
+import noop from 'lodash/noop';
+
+export interface AzureADOptions {
+  cluster: string;
+  config: Configuration;
+  debug?: boolean;
+}
+export const LOGIN_POPUP = 'loginPopup';
+export const LOGIN_REDIRECT = 'loginRedirect';
+export type AzureADSingInType = typeof LOGIN_POPUP | typeof LOGIN_REDIRECT;
+
+const loggerCallback = (level: LogLevel, message: string, containsPi: any) => {
+  if (containsPi) {
+    return;
+  }
+  switch (level) {
+    case LogLevel.Error:
+      console.error(message);
+      return;
+    case LogLevel.Info:
+      console.info(message);
+      return;
+    case LogLevel.Verbose:
+      console.debug(message);
+      return;
+    case LogLevel.Warning:
+      console.warn(message);
+  }
+};
+
+export class AzureAD {
+  private msalApplication: PublicClientApplication;
+  private account?: AccountInfo;
+  private userScopes = ['User.Read'];
+  private cluster: string = '';
+
+  static getDefaultMSALConfig(
+    debug: boolean = false
+  ): { cache: CacheOptions; system?: BrowserSystemOptions } {
+    return {
+      cache: {
+        cacheLocation: 'localStorage',
+        storeAuthStateInCookie: false,
+      },
+      system: debug ? { loggerOptions: { loggerCallback } } : undefined,
+    };
+  }
+
+  constructor({ cluster, config, debug }: AzureADOptions) {
+    const msalDefaultConfig = AzureAD.getDefaultMSALConfig(debug);
+    this.msalApplication = new PublicClientApplication({
+      ...msalDefaultConfig,
+      ...config,
+    });
+    this.setCluster(cluster);
+  }
+
+  setCluster(clusterName: string): void {
+    this.cluster = clusterName;
```

Comment:
That's true, with the next attempt to get a token or request api you'll pass through the `authentication()` step to get a valid token first, which you might expect to get when changing cluster. In case you find this redundant, we can remove it until somebody requests it. Then we might ask about expected flow in case of changing cluster